glTF improvements

moofMonkey opened the pull request:
This PR fixes
- Animations being duplicated in resulting glTF which resulted in huge glTF size - by reusing the same skeleton
- Animations related errors in resulting glTF caused by mixing `WithLocalTransform` and `WithLocalTranslation`/`WithLocalRotation`
- Int32 overflow check in WriteModelFile

and adds animation channel data deduplication.

Fixes most, if not all cases of #379.
Dota 2 `models/items/spectre/spectre_arcana/spectre_arcana_base.vmdl` mentioned in the issue gets perfectly exported with a more or less reasonable size of 75.5MB (instead of 1GB+).
![sandbox babylonjs com_](https://user-images.githubusercontent.com/11695747/206851569-89d31a5a-e193-47c9-9fb9-d57cd4a7872f.png)

moofMonkey commented:
I'll add logic to omit static data, as in my test it omits ~30% of resulting model size (126MB result)

Perryvw commented on ValveResourceFormat/IO/GltfModelExporter.cs:
```diff
@@ -346,7 +359,7 @@ public void ExportToFile(string resourceName, string fileName, VModel model)
                 if (string.IsNullOrEmpty(meshReference))
                 {
                     // If refmesh is null, take an embedded mesh
-                    meshes[i] = (embeddedMeshes[embeddedMeshIndex++], $"{modelName}.Embedded.{embeddedMeshIndex}");
```
Just checking: Does this not cause name collisions in files with multiple models with embedded meshes?

moofMonkey replied:
As far as I can tell, glTF doesn't really care about name collisions.

Either way, I removed `modelName` from embedded meshes names due to models now having their own nodes - this way there are less `WorldMatrix`es in resulting glTF, and it's less bloated when there are multiple models present.
https://github.com/SteamDatabase/ValveResourceFormat/pull/467/files#diff-a740f76f43bd153e9126ae3abe86209608d28825f381ed37b2d74aa3bcbc9c93R278
![image](https://user-images.githubusercontent.com/11695747/206853854-3559016d-0170-41e5-aecf-3a737771e63f.png)

moofMonkey commented:
181MB => 75.5MB. That's pretty nice, and keyframes should be identical.

xPaw commented on ValveResourceFormat/IO/GltfModelExporter.cs:
```diff
@@ -309,21 +266,110 @@ public void ExportToFile(string resourceName, string fileName, VModel model)
+
+                            if (!lastRotationDict.TryGetValue(bone, out var lastRotation) || lastRotation != boneFrame.Value.Angle)
+                            {
+                                if (rotationOmittedSet.Contains(bone))
```
Use Remove directly, it returns a bool.

moofMonkey commented:
Added comments for `rotationOmittedSet`/`translationOmittedSet` and fixed `GetGltfSkeletonJoints` logic for `numJoints`.

moofMonkey commented:
Fixed exception in case where Skeleton#Bones contained `null` for some reason - only applicable to this PR, as I removed recursion when retrieving joints.

Whole Dota 2 `models/creeps` folder gets exported without exceptions now.

moofMonkey commented:
Probably fixed exception when model didn't have meshes and skeletons - happened when I tried to export Dota 2 `maps/dota.vpk` `maps/dota/world.vwrld` as glTF.

Fixed standalone meshes not being transformed because WorldMatrix was set for `skeletonNode`, and not `modelNode`, and only models have `skeletonNode`.

Fixed LoD != 0 being exported to glTF.

xPaw commented:
> I tried to remove exporting of LoD != 0, but apparently it breaks export of some models like Dota 2 models/items/spectre/spectre_arcana/spectre_arcana_base.vmdl.

Does it break in the gui viewer too?

kristiker commented:
>I tried to remove exporting of LoD != 0

I was thinking of adding this extension: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/MSFT_lod. Looks like the engine that you're using supports it. Although an option to ignore them completely would be useful too.

moofMonkey commented:
> Does it break in the gui viewer too?

No, the GUI viewer loads meshes differently and I copied code from there, which is why it didn't work.
It was breaking because of `LoadModelMeshes` `string.IsNullOrEmpty(meshReference)` being incompatible with my first try. Basically embedded meshes didn't load at all.
Reimplemented `LoadModelMeshes` and now it doesn't export LoD != 0 and works properly.

> I was thinking of adding this extension: https://github.com/KhronosGroup/glTF/blob/main/extensions/2.0/Vendor/MSFT_lod. > Looks like the engine that you're using supports it.

At this time SharpGLTF [lacks support](https://github.com/vpenades/SharpGLTF/blob/master/src/SharpGLTF.Core/README.md#:~:text=MSFT_texture_dds-,MSFT_lod,-When%20this%20extension) for MSFT_lod.

kristiker commented:
It's not working right when performing a folder export like `dota/models/architecture`.

moofMonkey commented:
> It's not working right when performing a folder export like dota/models/architecture. Error:
> Failed to extract 'models/architecture/continuum/continuum_ground_01_crn_02.vmdl_c' - LogicalParent mismatch (Parameter 'value')

Sorry, can't reproduce errors on this specific file nor folder at 08c567ba69d0bd0365d665ab858be39276ef1171. What commit are you at?
Model is being exported without any errors, and properly opens at sandbox.babylonjs.com.
![sandbox babylonjs com_ (1)](https://user-images.githubusercontent.com/11695747/206870417-81d5e56b-431f-49a6-8eac-b35625abbea5.png)

> Single models though export so much faster now 💯. Yesterday this model pulled 5GB of ram to export, now it goes super quick!

Glad to hear that, hope it's easier to make your PR now.

xPaw commented:
I get these errors too when exporting a *folder*.

moofMonkey commented:
Can confirm, sorry, didn't check the console tab while exporting folders.

Will investigate and submit yet another PR if it doesn't get fixed before that, I guess.

moofMonkey commented:
The problem is in `LoadedUnskinnedMeshDictionary` being keyed by the mesh name, which was changed in this PR to not include model name for embedded meshes.

There won't be any problems with the Decompiler, but GUI for some reason reuses single instance of `GltfModelExporter`, and that causes meshes from different files to mix up. Guess this PR just revealed an existing problem.

kristiker commented:
Should clear the dictionary after every LoadModel then I think (and also on the Vmesh case that doesn't call LoadModel).
